lib/Backend/Vault/Kv/Internal.hs

```diff
@@ -0,0 +1,336 @@
+-- in 'KvResponse' at the 'data' key.
+-- For example:
+--
+-- { "keys":
+--   [ "a1"
+--   , "q1"
+--   ]
+-- }
+newtype ListSecrets = ListSecrets [T.Text]
+  deriving (Show)
+makeLenses ''ListSecrets
+
+-- |
+-- A type defining the response from the Vault server at endpoint
+-- '/v1/<MOUNT>/data/<PATH>?version=<VERSION>'. This is all actually always wrapped
+-- in 'KvResponse' at the 'data' key.
+-- For example:
+--
+-- { "data":
+--   { "token": "stuff"
+--   }
+-- , "metadata": <ReadSecretMetadata>
+--   { "created_time": "2018-03-22T02:24:06.945319214Z",
+--     "custom_metadata":
+--     { "owner": "jdoe",
+--       "mission_critical": "false"
+--     },
+--     "deletion_time": "",
+--     "destroyed": false,
+--     "version": 2
+--   }
+-- }
+data ReadSecret =
+  ReadSecret
+  { _secret :: HS.HashMap T.Text T.Text
+  , _rCustomMetadata :: HS.HashMap T.Text T.Text
+  , _createdTime :: T.Text
+  , _deletionTime :: T.Text
+  , _destroyed :: Bool
+  , _version :: Int
+  }
+  deriving (Show)
+makeLenses ''ReadSecret
+
+
+-- |
+-- A type defining the request to the Vault servet at endpoint
+-- '/v1/<MOUNT>/data/<PATH>'.
+-- For example:
+--
+-- { "options":
+-- , { "cas": 0
+--   }
+-- , "data":
+--   { "foo": "bar"
+--   , "zip": "zap"
+--   }
+-- }
+data PostSecret =
+  PostSecret
+  { _cas :: Maybe Int
+  , _pdata :: HS.HashMap T.Text T.Text
+  }
+  deriving (Show)
+type PatchSecret = PostSecret
+makeLenses ''PostSecret
+
+
+-- |
+-- A type defining the request to the Vault servet at endpoint
+-- '/v1/<MOUNT>/metadata/<PATH>'.
+-- For example:
+--
+-- { "max_versions": 5
+-- , "cas_required": false
+-- , "delete_version_after": "40m"
+-- , "custom_metadata":
+--   { "example": ":)"
+--   }
+-- }
+data UpdateMetadata =
+  UpdateMetadata
+  { _maxVersions :: Maybe Int
+  , _casRequired :: Maybe Bool
+  , _deleteVersionAfter :: Maybe T.Text
+  , _uCustomMetadata :: HS.HashMap T.Text T.Text
+  }
+  deriving (Show)
+makeLenses ''UpdateMetadata
+
+-- Overloaded Lens accessors
+
+class DataLens a b where
+  ddata :: Lens' a b
+
+instance a ~ b => DataLens (KvResponse a) b where
+  ddata = kdata
+
+instance DataLens PostSecret (HS.HashMap T.Text T.Text) where
+  ddata = pdata
+
+class CustomMetadataLens a where
+  customMetadata :: Lens' a (HS.HashMap T.Text T.Text)
+
+instance CustomMetadataLens ReadSecret where
+  customMetadata = rCustomMetadata
+
+instance CustomMetadataLens UpdateMetadata where
+  customMetadata = uCustomMetadata
+
+-- JSON serialization/deserialization, logically some ADTs need to only be deserialized,
+-- others serialized, but never both.
+
+instance A.FromJSON ListSecrets where
+  parseJSON = A.withObject "ListSecrets" $ \o ->
+    ListSecrets <$> o .: "keys"
+
+instance A.FromJSON ReadSecret where
+  parseJSON = A.withObject "ReadSecret" $ \o -> do
+    metadata <- o .: "metadata"
+    ReadSecret
+      <$> o .: "data"
+      <*> (metadata .: "custom_metadata" <|> pure (HS.fromList []))
```

Comment:
Done

Reply:
```
 <|> error "Cannot parse 'custom_metadata' field
```

No need for `<|>` here. Simply using `.:!` and `.!=` will ensure the entire `ReadSecret` parser fails if the `custom_metadata` key has an unexpected format.

Reply:
Also, I would use `.:?` instead of `.:!` here.

From the `.:?` docs:

> The result is Nothing if the key is not present or if its value is Null

From the `.:!` docs:

> This differs from .:? by attempting to parse Null the same as any other JSON value, instead of interpreting it as Nothing.

We want `"custom_metadata" : null` to be parsed as `_rsCustomMetadata = Nothing`, so `.:?` seems a better fit.

Reply:
Done